providers/gitlab.go

```diff
@@ -206,15 +287,70 @@ func (p *GitLabProvider) EnrichSession(ctx context.Context, s *sessions.SessionS
 		return fmt.Errorf("user email is not verified")
 	}
 
-	// Check group membership
-	// TODO (@NickMeves) - Refactor to Authorize
-	err = p.verifyGroupMembership(userInfo)
-	if err != nil {
-		return fmt.Errorf("group membership check failed: %v", err)
-	}
-
 	s.User = userInfo.Username
 	s.Email = userInfo.Email
 
+	p.addGroupsToSession(ctx, s)
+
+	p.addProjectsToSession(ctx, s)
+
 	return nil
+
+}
+
+// addGroupsToSession projects into session.Groups
+func (p *GitLabProvider) addGroupsToSession(ctx context.Context, s *sessions.SessionState) {
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, group := range p.Groups {
+		s.Groups = append(s.Groups, fmt.Sprintf("group:%s", group))
+	}
+
+}
+
+// addProjectsToSession projects into session.Groups
+func (p *GitLabProvider) addProjectsToSession(ctx context.Context, s *sessions.SessionState) {
+	// Iterate over projects, check if oauth2-proxy can get project information on behalf of the user
+	for _, project := range p.Projects {
+		projectInfo, err := p.getProjectInfo(ctx, s, project.Name)
+
+		if err == nil && !projectInfo.Archived {
+			// try first with project access
+			perms := projectInfo.Permissions.ProjectAccess
+			if perms == nil {
+				// use group project access as fallback
+				perms = projectInfo.Permissions.GroupAccess
+			}
+
+			if perms.AccessLevel >= project.AccessLevel {
+				s.Groups = append(s.Groups, fmt.Sprintf("project:%s", project.Name))
+			} else {
+				logger.Errorf("Warning: user %q does not have the minimum required access level for project %q", s.Email, project.Name)
+			}
+		}
+
+		if err != nil {
+			logger.Errorf("Warning: project info request failed: %v", err)
+		}
+
+		if projectInfo != nil && projectInfo.Archived {
```

Comment:
If this is a valid codepath (projectInfo = nil & err = nil). Then we are potentially calling a method on `nil` above:
on line 316: `if err == nil && !projectInfo.Archived {`

Reply:
It seems looking at `getProjectInfo`, `projectInfo` will only be nil if you had an error.

So I think if you start everything with the error handling, you can do the follow up checks with just `projectInfo.Archived`.

Ideally, this `if projectInfo.Archived` check can go first and you can `return` from it if it is archived.  Then we can reduce the conditional branching nesting that the main case was in and just have all that logic at the end of this method.